kafka.go

```diff
@@ -0,0 +1,390 @@
+
+		if publishBuilderEntry.traceId == "" {
+			publishBuilderEntry.traceId = generateID()
+		}
+
+		if publishBuilderEntry.messageId == "" {
+			publishBuilderEntry.messageId = generateID()
+		}
+
+		publishBuilderEntry.encoded, publishBuilderEntry.err = json.Marshal(Message{
+			Message:     publishBuilderEntry.message,
+			MessageType: publishBuilderEntry.messageType,
+			Service:     publishBuilderEntry.service,
+			TraceId:     publishBuilderEntry.traceId,
+			MessageId:   publishBuilderEntry.messageId,
+		})
+	}
+}
+
+// Encode PublishBuilder into array of bytes
+func (publishBuilderEntry *PublishBuilderEntry) Encode() ([]byte, error) {
+	publishBuilderEntry.ensureEncoded()
+	return publishBuilderEntry.encoded, publishBuilderEntry.err
+}
+
+// Length returns size of encoded value
+func (publishBuilderEntry *PublishBuilderEntry) Length() int {
+	publishBuilderEntry.ensureEncoded()
+	return len(publishBuilderEntry.encoded)
+}
+
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokerList []string) (*KafkaClient, error) {
+	config := sarama.NewConfig()
+	config.Version = sarama.V2_1_0_0
+
+	// currently only support 1 message broker
+	broker := sarama.NewBroker(brokerList[0])
+	_ = broker.Open(config)
+	if connected, err := broker.Connected(); !connected {
+		logrus.Error("Kafka not connected")
+		return nil, err
+	}
+
+	configAsync := sarama.NewConfig()
+
+	asyncProducer, err := sarama.NewAsyncProducer(brokerList, configAsync)
+	if err != nil {
+		logrus.Error("Unable to create async producer in kafka : ", err)
+		return nil, err
+	}
+
+	configSync := sarama.NewConfig()
+	configSync.Producer.Return.Successes = true
+	syncProducer, err := sarama.NewSyncProducer(brokerList, configSync)
+	if err != nil {
+		logrus.Error("Unable to create sync producer in kafka : ", err)
+		return nil, err
+	}
+
+	consumer, err := sarama.NewConsumer(brokerList, config)
+	if err != nil {
+		logrus.Error("Unable to create consumer in kafka : ", err)
+		return nil, err
+	}
+
+	client := &KafkaClient{
+		asyncProducer,
+		syncProducer,
+		consumer,
+		broker,
+	}
+
+	subscribeMap = make(map[string]map[string]func(message *Message, err error))
+
+	go listenAsyncError(asyncProducer)
+	go cleanup(client)
+	return client, nil
+}
+
+func cleanup(client *KafkaClient) {
+	// define signal notify
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+
+	func() {
+		for {
+			select {
+			case <-sig:
+				_ = client.asyncProducer.Close()
+				_ = client.syncProducer.Close()
+				_ = client.consumer.Close()
+				_ = client.broker.Close()
+			}
+		}
+	}()
+}
+
+func listenAsyncError(producer sarama.AsyncProducer) {
+	for err := range producer.Errors() {
+		logrus.Error("Unable to publish message using async producer to kafka : ", err)
+	}
+}
+
+// SetTimeout set listening timeout for publish with response
+func (client *KafkaClient) SetTimeout(timeout int) {
+	if timeout < 10000 {
```

Comment:
don't you think it's better to put this into a const?

Reply:
sure, I will update it
thanks